pubsub/kafkapubsub/kafka.go

```diff
@@ -0,0 +1,553 @@
+	// Clear the current session.
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.sess = nil
+	return nil
+}
+
+// ConsumeClaim implements sarama.ConsumerGroupHandler.ConsumeClaim.
+// This is where messages are actually delivered, via a channel.
+func (s *subscription) ConsumeClaim(sess sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
+	for msg := range claim.Messages() {
+		// We've received a message. Send it over the channel to ReceiveBatch.
+		// s.ch has no buffer, so the send blocks until ReceiveBatch is
+		// ready to receive them.
+		select {
+		case s.ch <- msg:
+		case <-sess.Context().Done():
+			// This session is over, we must return. We'll end up dropping msg, but
+			// that's OK.
+		}
+	}
+	return nil
+}
+
+// ReceiveBatch implements driver.Subscription.ReceiveBatch.
+func (s *subscription) ReceiveBatch(ctx context.Context, maxMessages int) ([]*driver.Message, error) {
+	// Try to read maxMessages for up to 100ms before giving up.
+	waitForMaxCtx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
+	defer cancel()
+
+	var dms []*driver.Message
+	for {
+		select {
+		case msg := <-s.ch:
+			// Read the metadata from msg.Headers.
+			md := map[string]string{}
+			for _, h := range msg.Headers {
+				md[string(h.Key)] = string(h.Value)
+			}
+			// Add a metadata entry for the message key if appropriate.
+			if len(msg.Key) > 0 && s.opts.KeyName != "" {
+				md[s.opts.KeyName] = string(msg.Key)
+			}
+			ack := &ackInfo{msg: msg}
+			dm := &driver.Message{
+				Body:     msg.Value,
+				Metadata: md,
+				AckID:    ack,
+				AsFunc: func(i interface{}) bool {
+					// TODO(rvangent): Implement.
+					return false
+				},
+			}
+			dms = append(dms, dm)
+
+			s.mu.Lock()
+			s.unacked = append(s.unacked, ack)
+			s.mu.Unlock()
+
+			if len(dms) == maxMessages {
+				return dms, nil
+			}
+		case <-s.closeCh:
+			// Fatal error, probably the topic doesn't exist.
+			return nil, s.closeErr
+		case <-waitForMaxCtx.Done():
+			// We've tried for a while to get maxMessages, but didn't get enough.
+			// Return whatever we have (may be empty).
+			return dms, nil
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		}
+	}
+}
+
+// SendAcks implements driver.Subscription.SendAcks.
+func (s *subscription) SendAcks(ctx context.Context, ids []driver.AckID) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// Mark them all acked.
+	for _, id := range ids {
+		id.(*ackInfo).acked = true
+	}
+	if s.sess == nil {
+		// We don't have a current session, so we can't send offset updates.
+		// We'll just wait until next time and retry.
+		return nil
+	}
+	// Mark all of the acked messages at the head of the slice. Since Kafka only
+	// stores a single offset, we can'tmark messages that aren't at the head; that
+	// would move the offset past other as-yet-unacked messages.
+	for len(s.unacked) > 0 && s.unacked[0].acked == true {
+		s.sess.MarkMessage(s.unacked[0].msg, "")
+		s.unacked = s.unacked[1:]
```

Comment:
FYI, I was a bit nervous about this pattern of appending to the end of a slice and then reslicing to drop the first entry; it wasn't immediately obvious to me that it would abandon the earlier slice entries correctly. It does: https://play.golang.org/p/6B1vAwoknLP.